subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/properties/annotations/NestedBeanAnnotationHandlerTest.groovy

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.properties.annotations
+
+import org.gradle.api.Action
+import spock.lang.Specification
+import spock.lang.Unroll
+
+class NestedBeanAnnotationHandlerTest extends Specification {
+    @Unroll
+    def "correct implementation for #type coerced to Action is tracked"() {
+        expect:
+        NestedBeanAnnotationHandler.getImplementationClass(implementation as Action) == implementation.getClass()
+
+        where:
+        type      | implementation
+        "Closure" | { it }
+        "Action"  |  new Action<String>() { @Override void execute(String s) {} }
+    }
+
+}
```

Comment:
Should we have some tests for discovering `@Nested` properties here?

Reply:
I added some unit tests. The are integration tests for the unit tests, so I am not sure if they add much value. WDYT?
Or do you have some unit tests in mind which I missed?